doc/design/codegen.md

```diff
@@ -0,0 +1,95 @@
+# SQLFlow Code Generator
+
+SQLFlow is a compiler that takes a SQL program and generates Argo workflow, which is a `.YAML` file,
+The Argo controller running on Kubernetes is the executor. This is a design doc on how to implement
+the back-end of the SQLFlow compiler.
+
+## The High-level Design of the Code Generator
+
+As mentioned above, SQLFlow compiler generates a `.YAML`, the piece code is as the following:
+
+``` yaml
+steps:
+    name: step-1
+    command: ["python", "-c"]
+    args: |
+        from runtime import tensorflow
+        tensorflow.train(....)
+    env:
+      name: SQLFLOW_OSS_AK
+      value: "xxxxxx"
+```
+
+From the above workflow `.YAML` file, each workflow step program contains three parts:
+
+1. The execution command as the `command` spec to execute the program.
+1. The execution program called it the submitter program, which could be written in Python, R, or Bash programming language.
+The submitter program submits an AI task on an AI platform e.g., Elastic, Alibaba PAI, or just running on host via `runtime` library.
+1. The runtime environment variables with the `env` spec.
+
+SQLFlow compiler provides the code generator component to generate the step program,
+the code generation is divided into the following stages:
+
+1. [Target Registry](#target-registry), registry a Code Generator in SQLFlow compiler.
+1. [CodeGenerator Interface](#code-generator-interface) is a Go interface that all code generators should implement it.
+1. [Code Emission](#code-emission) provides an assembler API to generate a step program.
+
+### Target Registry
+
+For a newly code generator, develops should registry it in SQLFlow compiler as the following pseudo-code:
+
+``` golang
+
+cgMapping = map[string]CodeGenerator {
+  "paiTensorFlow": PAITensorFlow{},
+  "paiXGBoost", PAIXGBoost{},
+  ...
+}
+```
+
+### Code Generator Interface
+
+For each code generator implementation, you should care about all IR types, different IR types have different behaviors and
+generate different submitter program. Each code generator owns an `ExecutionCtx` instance to tell Argo workflow
+on how to execute the target code.
+
+``` golang
+type ExecutionCtx struct {
+  ExecCommand []string      // How to execute the target code, .e.g ["python" "-c"]
+  Env map[string]string     // The environment variables for execution
+}
+
+type CodeGenerator interface {
+  ExecCtx() ExecutionCtx
+  Query(*ir.NormalStmt) (string, error)
+  Train(*ir.TrainStmt) (string, error)
+  Predict(*ir.PredictStmt) (string, error)
+  Explain(*ir.ExplainStmt) (string, error)
+  Evaluate(*ir.EvaluateStmt) (string, error)
+  ShowTrain(*ir.ShowTrainStmt) (string, error)
+  Optimize(*ir.OptimizeStmt) (string, error)
+  Run(*ir.RunStmt) (string, error)
+}
+```
+
+### Code Emission
```

Comment:
As the above comment, we may have to find a more accurate term.

Reply:
Code Emission -> Code Generation